addons/stock/models/stock_inventory.py

```diff
@@ -362,6 +367,38 @@ def _run_inventory_tasks(self, company_id=False):
         quants = self._get_neg_qty_quants(company_id)
         quants._create_or_update_neg_inv()
 
+        # conflicting lots check
+        domain = [('location_id.usage', 'in', ['internal', 'transit']), ('lot_id', '!=', False), ('product_id.tracking', '=', 'serial'), ('quantity', '!=', 0.0)]
+        if company_id:
+            domain = expression.AND([[('company_id', '=', company_id)], domain])
+        else:
+            domain = expression.AND([[('company_id', '!=', False)], domain])
+        quants = self.env['stock.quant'].search(domain)
+        lot_ids = quants.mapped('lot_id')
+        lot_conflicts = self.env['stock.production.lot']
+        for lot_id in lot_ids:
+            if len(quants.filtered(lambda q: q.lot_id == lot_id)) > 1:
```

Comment:
use a read_group with a group on lot_id and `count` as aggregate function. It will be way faster and use less lines :smiley: